src/hotspot/share/utilities/vmError.cpp

```diff
@@ -546,16 +615,41 @@ void VMError::report(outputStream* st, bool _verbose) {
 
 # define STEP(s) STEP_IF(s, true)
 
+# define REATTEMPT_STEP_IF_IMPL(s, cond, with_new_timeout) \
+    }                                                      \
+    _step_did_succeed = true;                              \
+  }                                                        \
+  if (_current_step < __LINE__ && !_step_did_succeed) {    \
+    _current_step = __LINE__;                              \
+    _current_step_info = s;                                \
+    if ((with_new_timeout)) {                              \
+      record_step_start_time();                            \
+      _step_did_timeout = false;                           \
+    }                                                      \
+    const bool cond_value = (cond);                        \
+    if (cond_value && should_stop_reattempt_step(          \
+                          stop_reattempt_reason)) {        \
+      st->print_cr("[stop reattempt (%s) reason: %s]",     \
+                   _current_step_info,                     \
+                   stop_reattempt_reason);                 \
+    } else if (cond_value) {
+      // [Continue Step logic]
+
+# define REATTEMPT_STEP_IF(s, cond)                        \
+  REATTEMPT_STEP_IF_IMPL(s, cond, false)
+
+# define REATTEMPT_STEP_WITH_NEW_TIMEOUT_IF(s, cond)       \
+  REATTEMPT_STEP_IF_IMPL(s, cond, true)
+
```

Comment:
I'm doubtful about the reset-timeout feature. If something timeouts, the chance is very high it will timeout again. Either because we have a deadlock, or because what we do is simply very slow. One example for very slow is printing callstacks - decoding debug info can be very slow if debug info is loaded e.g. from network share, but it will not get any faster by repeating the attempt.

With crashes related to printing registers and stack slots, I can see the sense and usefulness of reattempts. But timeouts are both more "sticky" (high chance of happening again) as well as worse than crashes. Customers want the crashing VM to be down quickly, to release all locks and files, so that the replacement VM can start up. 

So maybe we should scrap the new timeout feature. Would also simplify coding a bit.

Reply:
I removed it. 

Think I added it originally in the rework to not change the behaviour of the stack trace printing. But if it is as you say that if with source timeouts then without source is also likely to timeout, maybe they should share a timeout.